crates/cargo-util/src/registry.rs

```diff
@@ -11,11 +11,20 @@ pub fn make_dep_path(dep_name: &str, prefix_only: bool) -> String {
     } else {
         ("/", dep_name)
     };
-    match dep_name.len() {
+    match dep_name.chars().take(4).count() {
         1 => format!("1{}{}", slash, name),
         2 => format!("2{}{}", slash, name),
-        3 => format!("3/{}{}{}", &dep_name[..1], slash, name),
-        _ => format!("{}/{}{}{}", &dep_name[0..2], &dep_name[2..4], slash, name),
+        3 => {
+            let first_symbol = dep_name.chars().take(1).collect::<String>();
+
+            format!("3/{}{}{}", first_symbol, slash, name)
+        }
+        _ => {
+            let first_symbol = dep_name.chars().take(2).collect::<String>();
+            let second_symbol = dep_name.chars().skip(2).take(2).collect::<String>();
+
+            format!("{}/{}{}{}", first_symbol, second_symbol, slash, name)
+        }
```

Comment:
Could you find where the other half of this lives (where the directories are written to) so we make sure we match these up exactly?

Reply:
This function is used to get a path of folder of searched crate in crates-io index, so there is no the other half. cargo is only tries to read these directories.

Reply:
> crates-io index

This is the other half.

Reply:
As for now, non-ASCII crate name is not supported in both Cargo and crates.io[^1]. I am not sure whether any 3rd party registry works fine with that.

[^1]: https://github.com/rust-lang/crates.io/blob/2e744af8573f90b9706fa72ccffe5a52f91e2fdc/src/models/krate.rs#L267-L270